### PR TITLE
support SQS propagation

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?
Adds support for propagation through SQS MessageAttributes.

### Motivation
Users need propagation through SQS.

### Additional Notes
This only supports the case where there's only one message in `receiveMessage` response, which is the default.

Tests aren't currently running for `aws-sdk` in CircleCI, so either that needs to be fixed, or reviewers should run the tests locally.

### CODE REVIEW

File: packages/datadog-plugin-aws-sdk/test/index.spec.js [hunk 1]
```diff
@@ -311,6 +311,40 @@ describe('Plugin', () => {
             helpers.baseSpecs(semver, version, agent, getService, operation,
               serviceName, klass, sqsGetParams, key, metadata)
           })
+
+          describe('propagation', () => {
```
Comment (rochdev) [inline packages/datadog-plugin-aws-sdk/test/index.spec.js:315]: Why a `describe` for a single test without description?

File: packages/datadog-plugin-aws-sdk/test/index.spec.js [hunk 2]
```diff
@@ -297,7 +297,7 @@ describe('Plugin', () => {
             })
           })
 
-          after(done => {
+          afterEach(done => {
```
Comment (rochdev) [inline packages/datadog-plugin-aws-sdk/test/index.spec.js:300]: Why is this change needed?

File: packages/datadog-plugin-aws-sdk/test/fixtures/sqs.js [hunk 3]
```diff
@@ -5,7 +5,7 @@ const sqs = {}
 sqs['create'] = {
   QueueName: 'SQS_QUEUE_NAME',
   Attributes: {
-    'DelaySeconds': '60',
+    // 'DelaySeconds': '60',
```
Comment (rochdev) [inline packages/datadog-plugin-aws-sdk/test/fixtures/sqs.js:8]: Remove dead code.

File: packages/datadog-plugin-aws-sdk/src/services/sqs.js [hunk 4]
```diff
@@ -11,6 +11,38 @@ class Sqs {
...
+    if (operation === 'sendMessage') {
+      if (!request.params) {
+        request.params = {}
+      }
+      if (!request.params.MessageAttributes) {
+        request.params.MessageAttributes = {}
+      }
+      const ddInfo = {}
+      tracer.inject(span, 'text_map', ddInfo)
+      request.params.MessageAttributes._datadog = {
```
Comment (rochdev) [inline packages/datadog-plugin-aws-sdk/src/services/sqs.js:40]: This is inconsistent with how we usually propagate tracing context and may not be compatible with other languages. Why not simply use `MessageAttributes` as the carrier?

File: packages/datadog-plugin-aws-sdk/src/index.js [hunk 5]
```diff
@@ -29,18 +29,37 @@ function createWrapRequest (tracer, config) {
...
       analyticsSampler.sample(span, config.analytics)
 
+      awsHelpers.requestInject(span, this, serviceName, tracer)
+
+      const request = this
+
       return tracer.scope().activate(span, () => {
+        let boundCb
+        if (typeof cb === 'function') {
+          boundCb = function wrappedCb (err, resp) {
```
Comment (rochdev) [inline packages/datadog-plugin-aws-sdk/src/index.js:48]: This logic should be pushed to the service helper.

File: packages/datadog-plugin-aws-sdk/src/services/sqs.js [hunk 6]
```diff
@@ -11,6 +11,38 @@ class Sqs {
       'aws.sqs.queue_name': params.QueueName || params.QueueUrl
     })
   }
+
+  responseExtract (params, operation, response, tracer) {
+    if (operation === 'receiveMessage') {
+      if (
+        (!params.MaxNumberOfMessages || params.MaxNumberOfMessages === 1) &&
+        response &&
+        response.Messages &&
```
Comment (rochdev) [inline packages/datadog-plugin-aws-sdk/src/services/sqs.js:20]: The presence of a message should not change the behavior.

File: packages/datadog-plugin-aws-sdk/src/services/sqs.js (hunk 4 above)
Comment (bengl) [reply packages/datadog-plugin-aws-sdk/src/services/sqs.js:40]: `MessageAttributes` is not a simple key-value object. Keys are keys, but values are objects containing type information and also the value itself.

For example, for strings, the value is `{ DataType: 'String', StringValue: 'some string with some stuff in it' }`. This doesn't work with `text_map`. I suppose the `text_map` data could be extracted and transformed into individual properties here. Would that be preferable?

File: packages/datadog-plugin-aws-sdk/test/index.spec.js (hunk 2 above)
Comment (bengl) [reply packages/datadog-plugin-aws-sdk/test/index.spec.js:300]: It ... probably isn't.

File: packages/datadog-plugin-aws-sdk/test/index.spec.js (hunk 1 above)
Comment (bengl) [reply packages/datadog-plugin-aws-sdk/test/index.spec.js:315]: No particular reason. I'll switch it to a single `it`.

File: packages/datadog-plugin-aws-sdk/src/services/sqs.js (hunk 6 above)
Comment (bengl) [reply packages/datadog-plugin-aws-sdk/src/services/sqs.js:20]: What do you mean? `MessageAttributes` are store in `Message` so yeah we're going to need a `Message` to do anything here, no?

File: packages/datadog-plugin-aws-sdk/src/index.js (hunk 5 above)
Comment (bengl) [reply packages/datadog-plugin-aws-sdk/src/index.js:48]: Hmm, alright I'll give that a shot.

File: packages/datadog-plugin-aws-sdk/src/services/sqs.js (hunk 6 above)
Comment (rochdev) [reply packages/datadog-plugin-aws-sdk/src/services/sqs.js:20]: What I mean is that the current code will fallback to the old behavior instead of `null` when there are no messages. Although I'm also not sure how there could be no messages.

Comment (bengl) [reply packages/datadog-plugin-aws-sdk/src/services/sqs.js:20]: > What I mean is that the current code will fallback to the old behavior instead of null when there are no messages.

That's what we want, isn't it?

> Although I'm also not sure how there could be no messages.

When there are no messages in the queue, and you call `receiveMessage`, the `Messages` property will not be there at all, rather than being an empty array. At least, this is how localstack behaves.

Comment (rochdev) [reply packages/datadog-plugin-aws-sdk/src/services/sqs.js:20]: > That's what we want, isn't it?

It should explicitly set the parent to `null` to preserve the behavior for when propagation is supported.

> When there are no messages in the queue, and you call receiveMessage, the Messages property will not be there at all, rather than being an empty array. At least, this is how localstack behaves.

Should there even be a span at all in that case? I'm not sure how to deal with pulling since technically the "no message" case could not exist in a push model.

Comment (bengl) [reply packages/datadog-plugin-aws-sdk/src/services/sqs.js:20]: > It should explicitly set the parent to null to preserve the behavior for when propagation is supported.

There's no new span to set the parent on. See below.

> Should there even be a span at all in that case? I'm not sure how to deal with pulling since technically the "no message" case could not exist in a push model.

Right, in that case there should not be a span at all, and instead we just propagate scope. See [this if-block](https://github.com/DataDog/dd-trace-js/pull/963/commits/e15b5371d0dd896306e8d73d074604b3ca041a30#diff-5477a55323f98152ec99122e4240cd4bR76-R84).

File: ext/formats.js [hunk 7]
```diff
@@ -6,5 +6,6 @@ module.exports = {
   TEXT_MAP: opentracing.FORMAT_TEXT_MAP,
   HTTP_HEADERS: opentracing.FORMAT_HTTP_HEADERS,
   BINARY: opentracing.FORMAT_BINARY,
-  LOG: 'log'
+  LOG: 'log',
+  W3C: 'w3c'
```
Comment (rochdev) [inline ext/formats.js]: The W3C describes multiple formats, so it shouldn't itself be a format.
